File: app/services/vlc_player.rb
# Manages VLC playback via its HTTP interface.
#
# Reuses a running VLC instance when possible by sending commands through
# the HTTP interface.  Only spawns a new process when VLC is not already
# running.
#
# VLC HTTP interface: http://localhost:VLC_PORT/requests/status.json
#
class VlcPlayer
  VLC_PATH = ENV.fetch('VLC_PATH', '/Applications/VLC.app/Contents/MacOS/VLC')
  VLC_HTTP_PORT = ENV.fetch('VLC_HTTP_PORT', '9090').to_i
  VLC_HTTP_PASSWORD = ENV.fetch('VLC_HTTP_PASSWORD', 'simpsons')

  class << self
    # Play a file in VLC. Reuses the existing instance if one is running,
    # otherwise spawns a new one.
    def play(file_path, start_time: nil)
      if running?
        enqueue_and_play(file_path, start_time: start_time)
      else
        launch(file_path, start_time: start_time)
      end
    end

    # Stop playback (but keep VLC open).
    def stop
      request('?command=pl_stop')
    rescue StandardError
      nil
    end

    # Query VLC's HTTP interface for current playback status.
    # Returns a hash with :state, :time, :length, :position
    def status
      response = request('')
      return nil unless response

      {
        state: response['state'],
        time: response['time']&.to_i || 0,
        length: response['length']&.to_i || 0,
        position: response['position']&.to_f || 0.0
      }
    rescue StandardError => e
      Rails.logger.debug("VlcPlayer: status query failed — #{e.message}")
      nil
    end

    # Check if VLC is currently playing or paused.
    def active?
      s = status
      s && %w[playing paused].include?(s[:state])
    rescue StandardError
      false
    end

    # Check if VLC's HTTP interface is reachable.
    def running?
      status.present?
    rescue StandardError
      false
    end

    private

    # Launch a fresh VLC process with the HTTP interface enabled.
    def launch(file_path, start_time: nil)
      unless File.exist?(VLC_PATH)
        Rails.logger.warn("VlcPlayer: VLC not found at #{VLC_PATH}")
        return nil
      end

      args = [
        VLC_PATH,
        file_path,
        '--extraintf', 'http',
        '--http-port', VLC_HTTP_PORT.to_s,
        '--http-password', VLC_HTTP_PASSWORD,
        '--no-http-forward-cookies',
        '--one-instance'
      ]

      args += ['--start-time', start_time.to_s] if start_time && start_time.to_i > 0

      pid = spawn(*args, %i[out err] => '/dev/null')
      Process.detach(pid)

      Rails.logger.info("VlcPlayer: launched VLC (PID=#{pid}) for #{File.basename(file_path)}")
      pid
    end

    # Tell the running VLC instance to play a new file via the HTTP API.
    # Clears the playlist, adds the new file, and plays it.
    def enqueue_and_play(file_path, start_time: nil)
      encoded_path = URI.encode_www_form_component(file_path)

      # Clear playlist and add the new file
      request('?command=pl_empty')
      request("?command=in_play&input=#{encoded_path}")

      # Wait briefly for VLC to start loading the file, then seek if needed
      if start_time && start_time.to_i > 0
        sleep(0.5)
        request("?command=seek&val=#{start_time.to_i}")
      end

      Rails.logger.info("VlcPlayer: sent #{File.basename(file_path)} to running VLC instance")
      true
    end

    def request(query_path)
      require 'net/http'
      require 'json'

      uri = URI("http://localhost:#{VLC_HTTP_PORT}/requests/status.json#{query_path}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.open_timeout = 1
      http.read_timeout = 2

      req = Net::HTTP::Get.new(uri)
      req.basic_auth('', VLC_HTTP_PASSWORD)

      response = http.request(req)
      return nil unless response.is_a?(Net::HTTPSuccess)

      JSON.parse(response.body)
    end
  end
end

File: electron/services/vlc-player.js
// VLC playback control via HTTP interface on localhost:9090.
// Reuses running VLC instance; spawns new one if not running.

const { spawn } = require('child_process')
const path = require('path')
const fs = require('fs')
const crypto = require('crypto')

const VLC_PATH = process.env.VLC_PATH || '/Applications/VLC.app/Contents/MacOS/VLC'
const VLC_HTTP_PORT = parseInt(process.env.VLC_HTTP_PORT || '9090', 10)
// Generate a random password per app session to prevent local cross-app access
const VLC_HTTP_PASSWORD = process.env.VLC_HTTP_PASSWORD || crypto.randomBytes(16).toString('hex')

const AUTH_HEADER = 'Basic ' + Buffer.from(`:${VLC_HTTP_PASSWORD}`).toString('base64')

async function vlcRequest(queryPath = '') {
  const url = `http://127.0.0.1:${VLC_HTTP_PORT}/requests/status.json${queryPath}`
  try {
    const res = await fetch(url, {
      headers: { Authorization: AUTH_HEADER },
      signal: AbortSignal.timeout(3000),
    })
    if (!res.ok) return null
    return await res.json()
  } catch {
    return null
  }
}

async function status() {
  const data = await vlcRequest()
  if (!data) return null
  return {
    state: data.state,
    time: parseInt(data.time) || 0,
    length: parseInt(data.length) || 0,
    position: parseFloat(data.position) || 0,
  }
}

async function isRunning() {
  const s = await status()
  return s != null
}

async function isActive() {
  const s = await status()
  return s != null && (s.state === 'playing' || s.state === 'paused')
}

function launchVlc(filePath, startTime) {
  if (!fs.existsSync(VLC_PATH)) {
    console.warn(`VlcPlayer: VLC not found at ${VLC_PATH}`)
    return null
  }

  const args = [
    filePath,
    '--extraintf', 'http',
    '--http-host', '127.0.0.1',
    '--http-port', String(VLC_HTTP_PORT),
    '--http-password', VLC_HTTP_PASSWORD,
    '--no-http-forward-cookies',
  ]

  if (startTime && parseInt(startTime) > 0) {
    args.push('--start-time', String(parseInt(startTime)))
  }

  const child = spawn(VLC_PATH, args, {
    detached: true,
    stdio: 'ignore',
  })
  child.unref()
  console.log(`VlcPlayer: launched VLC (PID=${child.pid}) for ${path.basename(filePath)}`)
  return child.pid
}

async function enqueueAndPlay(filePath, startTime) {
  // Build a proper file:// URI — encode each path component individually
  // to correctly handle spaces, #, ?, and other special characters in filenames
  const fileUri = 'file://' + filePath.split('/').map(c => encodeURIComponent(c)).join('/')

  await vlcRequest('?command=pl_empty')
  await vlcRequest(`?command=in_play&input=${fileUri}`)

  if (startTime && parseInt(startTime) > 0) {
    await new Promise(r => setTimeout(r, 500))
    await vlcRequest(`?command=seek&val=${parseInt(startTime)}`)
  }

  console.log(`VlcPlayer: sent ${path.basename(filePath)} to running VLC instance`)
  return true
}

async function play(filePath, startTime = 0) {
  if (await isRunning()) {
    return enqueueAndPlay(filePath, startTime)
  }
  return launchVlc(filePath, startTime)
}

async function stop() {
  try {
    await vlcRequest('?command=pl_stop')
  } catch { /* ignore */ }
}

module.exports = { play, stop, status, isRunning, isActive }
